### DISCUSSION TITLE
Task exited with return code Negsignal.SIGSEGV when dag triggered from UI

### DISCUSSION THREAD

**Discussion by shelly2904** [apache/airflow#24463] [Q&A]
Hi, 

I am trying to trigger a DAG from Airflow UI which internally has a task to call requests.get(). This results in below error:
  `Task exited with return code Negsignal.SIGSEGV` 

This is working fine when tested using command `airflow dags test <dag_id> <execution_start_date>`

Below versions are used:
1. Airflow==2.3.1
2. Python==3.8.0
3. requests==2.27.1

Can you please advise?

Thanks.

**Comment by potiuk** [comment]
This is not an airflow issue. It's issue with your code.

You need to look at your code and see which component fails with sigsegv. It's  some of your code and custom libraries, sigsegv comes from C-libraries usually. You need to look how your deployment is done and try to reproduce it there. When you run stuff manually (from terminal) you have different environment than when airflow is run as a "daemon" or service (for example airflow has no terminal with user). You need to check how your airflow is deployed (possibly with your admin/devops) and analyse logs/core dumps produced on the server.  Usually (depends on your server configuration) when a process dies with sigsegv it produces "core dumps" - and you can trace them, but how and where - this is something that only you and your admin can track.

Again - not airflow issue, look at your deployment. Good luck.

**Reply by maver1ck** [reply]
@potiuk 
It's issue connected with using `os.fork`. Do you know if the Airflow is using it?
Reference: https://github.com/python/cpython/issues/105912

**Reply by potiuk** [reply]
Yes. A lot. and some of the usages are configurable.  But look through the whole thread here for various ways usage of some libraries might trigger SIGSEGV. You wil find many ways that you might try to mitigate your issue depending on the library you use. Unfortunately - (despite the 26 thumbs down that do not bring any value) this does not change the fact that Airflow on its own cannot make it any clearer on which library and usage of yours causes the problem.

**Reply by Taragolis** [reply]
Started from 2.5.1 we even write message if SIGSEGV happen in LocalTaskJobRunner how to debug

https://github.com/apache/airflow/blob/3c0a714cb57894b0816bf39079e29d79ea0b1d0a/airflow/jobs/local_task_job_runner.py#L47-L74

**Comment by yakimzhang** [comment]
the same error on mac

**Comment by shelly2904** [comment]
Hi,

I was trying to use airflow in MacOS and was getting the same error.
On debugging, I found out that it was crashing in urllib. Below is the bug

https://bugs.python.org/issue28342

I fixed it by adding an environmental variable NO_PROXY to bypass the
required host.

Let me know if this is helpful.

Thanks.

**Reply by yakimzhang** [reply]
os.environ['NO_PROXY'] = ' URL'    thanks

**Reply by binary-signal** [reply]
Thank you, I was running in the same issue when running a dev instance of airflow on macosx, your recommendation adding the NO_PROXY env variable resolved the issue  , kinda hard to figure out what was wrong since log messages even on debug mode on airflow didn't give lots of information besides a cryptic "Task exited with return code Negsignal.SIGSEGV"

**Reply by potiuk** [reply]
Unfortunately, this is the nature of SIGSEGV. When any library crashes HARD with segmentation violation, not only Airlfow has no chance to know the reason but even there is no slightest indication of which library caused it.

Luckily we have now this dicsussion thread that might help users like you to find possible reason for that. 

This is actually the ONLY way to help our users.

**Reply by potiuk** [reply]
But @binary-signal - if you can figure out some kind of magic to make the message less-criptic, I am all ears :D

**Reply by binary-signal** [reply]
@potiuk I am not sure about magic, although more debug messages would help to know what was the last code executed before the crash.

Strangely I had the issue only on apple m1 hardware, on apple intel hardware I had no problem  same OS same python version same dependencies go figure

**Reply by potiuk** [reply]
> @potiuk I am not sure about magic, although more debug messages would help to know what was the last code executed before the crash.

If you know how to do it - by all means please do it. I think it is impossible, but if you can prove that I am wrong, I am more than happy to accept PR for that.

**Reply by potiuk** [reply]
> Strangely I had the issue only on apple m1 hardware, on apple intel hardware I had no problem same OS same python version same dependencies go figure

This is pretty obvious. SIGSEGV can only be generated when low-level c-code crahes. Crashes in the way that whole Python interpreter crashses and this is why we cannot do anything about. And it can be caused by some depdency that has some low-level c-code that - for example - uses wrong processor instructions, or uses memory in a bad way. So it is completely not a surprise that it only crashes with a single architecture. This is pretty basic property of SIGSEGV, In case you were not aware of it.

Which dos not make it any easier to provide any more debugging around it, but if you think you know better, I am all ears.

**Reply by binary-signal** [reply]
It's also obvious before the SIGSEGV error a bunch of code was executed but even in debug mode there's no clue what is been executed in the log messages before the crash which doesn't leaves many options how to debug it in the first place.

**Reply by potiuk** [reply]
The logs you see is want had been executed.  Again. If you have an idea how to fix it - do it. Otherwise it's just a noise.

**Reply by binary-signal** [reply]
There were no other logs besides the SIGSEGV in debug mode

**Reply by potiuk** [reply]
What is your proposal? Still waiting for PR and continue getting useless noise.

**Reply by potiuk** [reply]
CC: @binary-signal -> as you might see there are people who (contrary to producing noise) can provide useful insights and generate some ideas that might turn into possible ways of circumventing the problem.

I highly recommend constructive approach and trying to be creative rather than continuous, useless complaining on things that cannot be fixed and are outside of our control. This is far better approach overall and one that helps to make world a better place.

**Reply by pingzh** [reply]
@shelly2904 saved my life. thanks

**Reply by lugonthier** [reply]
> @potiuk I am not sure about magic, although more debug messages would help to know what was the last code executed before the crash.
> 
> Strangely I had the issue only on apple m1 hardware, on apple intel hardware I had no problem same OS same python version same dependencies go figure

I had the exact same behaviour. I had the issue only on apple M1 but not on other hardware. thanks

**Reply by derrickqin** [reply]
Thanks, @shelly2904 Had the same issue, adding environment variable worked. `no_proxy=* airflow standalone`

**Reply by loupeipei** [reply]
Can i ask how to add this environmental variable NO_PROXY to mac ?
Much appreciated!

**Reply by Reindert94** [reply]
> Can i ask how to add this environmental variable NO_PROXY to mac ? Much appreciated!

```
import os

@task(retries=0)
def some_task():
    # bypass M1 Error
    os.environ['NO_PROXY'] = "URL"
```

**Reply by chris97425** [reply]
Thx a lot, work for me.

**Reply by keiraaa-xrq** [reply]
Hi thank you so much for this!! I searched at a lot of places and could not identify the issue until I saw this post. I am quite new to requests and airflow. Do you mind explaining what's the exact issue and why setting the env variable helps? Thanks in advance!

**Reply by potiuk** [reply]
I believe this is an issue with some incompatibilities and racing conditions of basic SSL libraries in Python on ARM simply c-generated code on ARM has a bug that triggers the issue when looking for proxy (i believe). This has nothing to do with airflow or even requests i think. You can find corresponding issues in Python. (just look it up in Google) for NO_PROXY and SIGSEGV and you will find relevant discussions if you are curious and want to dig deeper.

If you find more details that cna shed some light - please bring it back here if you find more exact and correct explanation. I would be courious to learn but had no time to dig deeper but f someeone has the curiosity and time i will be super happy to learn more.

**Reply by csm10495** [reply]
It's not just ARM. It can happen on amd64 too.

**Reply by potiuk** [reply]
This is the first time I hear it that it can happen on AMD with relation to NO_PROXY at least. It is likely another issue, but since there are no details, I am not 
 aware about NO_PROXY fix in relation to AMD.


Let me just say that SIGSEGV is completely independent from airflow because airflow is python program. You can develop any c-binary that produces SIGSEGV with two lines of code, so generally any binary that is executed might generate it if it has a bug. There might be as many sigsegv as many differnt binary libraries or binary processes you run (or more). Each of them might be different and each of them caused by different bug. And with the nature of sigsegv stopping an interpreter pretty much immediately, there is usually not much information airflow can provide about the root cause.

We discussed it extensively and in recent version of airflow you should get some hints on how you can debug it and try to find out which of the binaries cause it by setting PYTHONFAULTHANDLER ENV variable https://docs.python.org/3/library/faulthandler.html.
But basically eeach SIGSEGV might have a different reason and is pretty much always external to Airflow.

**Reply by csm10495** [reply]
The solution here: https://github.com/apache/airflow/discussions/24463#discussioncomment-4404542 was how we worked around it regardless of arch. We definitely were hitting it on amd64. 

I know it's not an airflow specific problem, it just so happens to get hit with airflow. This thread is very useful for folks who hit it in that case.

**Reply by potiuk** [reply]
Ah. OK. It's the first time I 've heard it is AMD as well. Thanks for letting me know @csm10495 )

I did some digging in and it does seem like a limitation of scproxy library not working correctly with MacOS  and likely it traces back to an old error that is closed as "3r-party bug". 

https://bugs.python.org/issue13829

And looks like more details are @keiraaa-xrq):

> 1) The crash in _scproxy is due to limitations in Apple's libraries, in particular they don't work in child processes created with os.fork() without calling execv*()

**Comment by yakimzhang** [comment]
from web3 import Web3
web3.eth.get_balance(address)

This code also triggers an Negsignal.SIGSEGV ，But I don't know how to fix yet

**Comment by chajath** [comment]
I also ran into this kind of issue when locally running the DAG that uses `s3fs`. Looking at the trace, I think root cause of a lot of issues reported here can be traced to issues like post-fork incompatibility on mac: https://github.com/python/cpython/issues/58037
It makes sense that these issues don't pop up when running with `dag test` but it is shown when running with a serial executor, as the executor forks before running the task.

To mitigate the issue, one easy solution could be to use a real simple executor that does not fork. Will see if there is existing Airflow setup that can achieve this through configuration.

**Reply by potiuk** [reply]
Actualy this is a good idea - We have it https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#execute-tasks-new-python-interpreter  and maybe it could be set to True by default for MacOS ? 

@chajath maybe you care for a PR implementing it ?

**Reply by potiuk** [reply]
If not I can do it, but it would be a nice way to become contributor :)

**Reply by chajath** [reply]
Great. I'll take a stab at it sometime this week when I get the chance to.

**Reply by leonardara** [reply]
@potiuk  i run into this issue with the `GCSFileTransformOperator`, but seems even after i set the env var `AIRFLOW__CORE__EXECUTE_TASKS_NEW_PYTHON_INTERPRETER` to `True` the issue still persists, how can i get around of it?

**Reply by potiuk** [reply]
No idea. There must be other reasons - as mentioned before, we are not able to track SIGSEGV source as it is independent from Airflow. Wild guess is that if you are on M1  and have older Python version/SSL people have (not only with Airflow - it's a general issue) similar problems. look for SIGSEGV in airflow issues/discussions and you will find some recent tasks that refer to it.

**Comment by herrykarmito92** [comment]
I'm facing same problem like on this forum. 
Thanks to @potiuk , mentions about library broken and nothing do with airflows.

I figure it out, last week I have a POC with telegram bot and installed a ton of Library package on my airflow virtual env.

Than I tried to so many tools to figure it out , and questioning my self : 
Q : How can i know which library using `pip freeze` or `pip list` is fresh installed  7 days from now ?
A : Lucky, I found this tools : [pip-date](https://github.com/E3V3A/pip-date) and try to uninstall some error library after using this tools scanning and run the airflow. 

After re-run airflow server again , airflow show error about `no module`.
Because I save the command before I run it in my notepad , then I can re-track again which library I need to installed again.

Note : 
- every version of `ariflow` needed very specific version of library

**Comment by csm10495** [comment]
I started hitting this when using requests with Airflow 2.5.0 on Mac.

We seem to be able to workaround it by running:

```
    from _scproxy import _get_proxy_settings

    _get_proxy_settings()
```

in the process that kicks off Airflow's scheduler, workers, etc.

Not a great solution.. but leaving here in case it helps someone else.

**Reply by Noushath** [reply]
> I started hitting this when using requests with Airflow 2.5.0 on Mac.
> 
> We seem to be able to workaround it by running:
> 
> ```
>     from _scproxy import _get_proxy_settings
> 
>     _get_proxy_settings()
> ```
> 
> in the process that kicks off Airflow's scheduler, workers, etc.
> 
> Not a great solution.. but leaving here in case it helps someone else.

True Saviour!!!
 **For mac airflow Task exited with return code Negsignal.SIGSEGV error** adding below to dag definition file saved my day with Airflow 2.5.0
from _scproxy import _get_proxy_settings

_get_proxy_settings()

Please find my configs below:

`Apache Airflow
version                | 2.5.0                                              
executor               | SequentialExecutor                                                                            
                                                                            

System info
OS              | Mac OS                                                                                                                                                             
architecture    | arm                                                                                                                                                                
uname           | uname_result(system='Darwin', node='US-P2QH240FMG', release='22.2.0', version='Darwin Kernel Version 22.2.0: Fri Nov 11 02:03:51 PST 2022;                         
                | root:xnu-8792.61.2~4/RELEASE_ARM64_T6000', machine='arm64')                                                                                                        
locale          | ('en_GB', 'UTF-8')                                                                                                                                                 
python_version  | 3.10.9 (main, Dec 15 2022, 17:11:09) [Clang 14.0.0 (clang-1400.0.29.202)]                                                                                          
            `

**Reply by htrueman** [reply]
You helped A LOT. Many thanks!

**Comment by RusabKhan** [comment]
This problem still exists on macOS.

Details:
macOS: Sonoma 14.4
Python: 3.9.18
Airflow: 2.8.1

A workaround is to add `os.environ['NO_PROXY'] = '*'` to your airflow dags.

**Reply by potiuk** [reply]
Yes. This is problem with Python not Airflow, so yep. It still exists.

**Comment by EricGao888** [comment]
Just encountered with this issue yesterday and spent half a day searching clues until reading this discussion.

The issue appeared when remote task handler trying to fetching logs from remote storage. Error output like:

```text
[2024-05-15 14:22:38 +0800] [54502] [ERROR] Worker (pid:54823) was sent SIGSEGV!
[2024-05-15 14:22:38 +0800] [54827] [INFO] Booting worker with pid: 54827
```

After setting env vars `no_proxy=*`, things worked out.

macOS Ventura 13.1
Python 3.10

Looks like not an Airflow problem. I record the log here just in case other users / developers come into the same issue when using remote logging.
